R/desirability.R

```diff
@@ -0,0 +1,228 @@
+  vals_chr <- purrr::map_chr(vals, ~ rlang::expr_deparse(.x))
+
+  is_good <- fns %in% vals
+  if (any(!is_good)) {
+    cli::cli_abort(
+      "The following functions are unknown to the  {.pkg desirability2} package:
+		{.fn {fns[!is_good]}}. Please use one of: {.or {.fn {all_f}}}.",
+      call = call
+    )
+  }
+  invisible(TRUE)
+}
+
+get_nms <- function(x) {
+  res <- names(x)
+  if (is.null(res)) {
+    res <- rep("", length(x))
+  }
+  # first name is for function
+  res[-1]
+}
+
+get_num_nms <- function(x) {
+  if (is.null(x)) {
+    res <- 0
+  } else {
+    res <- length(x)
+  }
+  res
+}
+
+check_arg_names <- function(x, call = rlang::env_parent()) {
+  num_args <- purrr::map_int(x, length) - 1L
+
+  if (any(num_args == 0)) {
+    cli::cli_abort(
+      "There needs to be at least one argument to the optimization goal
+      functions.",
+      call = call
+    )
+  }
+
+  nms <- purrr::map(x, get_nms)
+  num_nms <- purrr::map_int(nms, get_num_nms)
+  good_first_nm <- purrr::map_lgl(nms, ~ .x[1] == "")
+
+  if (any(!good_first_nm)) {
+    bad_first_nms <- which(!good_first_nm)
+    num_bad_first_nms <- length(bad_first_nms)
+    cli::cli_abort(
+      "{num_bad_first_nms} optimization goal{?s} {?has/have} a first argument that
+      is named. {?It/They} should be unnamed.",
+      call = call
+    )
+  }
+  good_other_names <- purrr::map_lgl(nms, ~ all(.x[-1] != ""))
+  if (any(!good_other_names)) {
+    bad_other_nms <- which(!good_other_names)
+    num_bad_other_nms <- length(bad_other_nms)
+    cli::cli_abort(
+      "{num_bad_other_nms} optimization goal{?s} {?has/have} additional arguments that
+      are {.strong not} named. All but the first argument should be named.",
+      call = call
+    )
+  }
+
+  invisible(TRUE)
+}
+
+# ------------------------------------------------------------------------------
+# Helper functions
+
+match_fn <- function(orig, vals) {
+  which(purrr::map_lgl(vals, ~ identical(.x, orig)))
+}
+
+index_fn <- function(fns, vals) {
+  purrr::map_int(fns, ~ match_fn(.x, vals = vals))
+}
+
+sub_fn <- function(x, ind, vals) {
+  x[[1]] <- vals[[ind]]
+  x
+}
+
+translate_fn_args <- function(x, vals, subs, .use_data) {
+  fns <- purrr::map(x, ~ .x[[1]])
+  ind <- index_fn(fns, vals = vals)
+
+  y <- purrr::map2(x, ind, ~ sub_fn(.x, .y, vals = subs))
+
+  if (.use_data) {
+    # In case the user did not specify all of the required arguments, tell the
+    # functions to use the data to impute them. If they did set values for all
+    # of the required arguments, `use_data = TRUE` has no effect.
+    y <- purrr::map(y, ~ rlang::call_modify(.x, use_data = TRUE))
+  }
+  y
+}
+
+# ------------------------------------------------------------------------------
+# Tools to help determine if we need to splice in known args (like low or high)
+# or to add `use_data = TRUE` to fil in the blanks with the data.
+
+get_missing_args <- function(x) {
```

Comment:
I'm not using these right now but they (or something like them) would probably be helpful for @franceslinyc